### PR TITLE
ci: Use busybox utils for one build

### PR DESCRIPTION
**Pull request by maflcko**
To make sure Bitcoin Core can be built with BusyBox, see https://github.com/bitcoin/bitcoin/pull/16927#issuecomment-536483706

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Comment by practicalswift** [comment]
Concept ACK

**Comment by dongcarl** [comment]
@MarcoFalke I'm guessing there's a reason why we don't wanna add another build?

**Comment by MarcoFalke** [comment]
Yeah, another build means the ci is taking longer. I think using BusyBox on one build that bootstraps from `./depends` (with gui and wallet) is sufficient.

File: ci/test/04_install.sh [hunk 1]
```diff
@@ -33,18 +33,28 @@ if [ -z "$RUN_CI_ON_HOST" ]; then
...
+DOCKER_EXEC echo "Number of CPUs \(nproc\):" \$\(nproc\)
 
 ${CI_RETRY_EXE} DOCKER_EXEC apt-get update
 ${CI_RETRY_EXE} DOCKER_EXEC apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES $DOCKER_PACKAGES
 
+if [ "$USE_BUSY_BOX" = "true" ]; then
+  echo "Setup to use BusyBox utils"
+  DOCKER_EXEC mkdir -p $BASE_SCRATCH_DIR/bins/
+  # tar excluded for now because it requires passing in the exact archive type in ./depends (fixed in later BusyBox version)
+  # find excluded for now because it does not recognize the -delete option in ./depends (fixed in later BusyBox version)
```
Comment (fanquake) [inline ci/test/04_install.sh]: Support for `-delete` was added to `find` in BusyBox `1.6.0` (2007). The [following](https://github.com/bitcoin/bitcoin/blob/master/depends/packages/xcb_proto.mk#L25) inside an Alpine container (BusyBox v1.30.1) seems to work ok:
```bash
touch something.pyo something.pyc
find -name "*.pyc" -delete && find -name "*.pyo" -delete
```

What error were you seeing?

File: ci/test/04_install.sh [hunk 2]
```diff
@@ -33,18 +33,28 @@ if [ -z "$RUN_CI_ON_HOST" ]; then
...
+  # ar excluded for now because it does not recognize the -q option in ./depends (unknown if fixed)
+  # shellcheck disable=SC1010
+  DOCKER_EXEC for util in \$\(busybox --list \| grep -v "^ar$" \| grep -v "^tar$" \| grep -v "^find$"\)\; do ln -s \$\(command -v busybox\) $BASE_SCRATCH_DIR/bins/\$util\; done
+  DOCKER_EXEC patch --help
```
Comment (fanquake) [inline ci/test/04_install.sh]: Assume we are doing this to get the BusyBox version printed? I see `BusyBox v1.27.2 (Ubuntu 1:1.27.2-2ubuntu3.2)`

**Review by fanquake — COMMENTED** [review]
Concept ACK - am testing locally.

File: ci/test/04_install.sh (hunk 1 above)
Comment (laanwj) [reply ci/test/04_install.sh]: 2007 oh my, please don't tell me anything that old is being used :fearful:

File: ci/test/04_install.sh (hunk 2 above)
Comment (MarcoFalke) [reply ci/test/04_install.sh]: Yes, fixed in commit ddddd

**Comment by MarcoFalke** [comment]
See here: https://travis-ci.org/MarcoFalke/bitcoin-core/jobs/592530914#L4480

And:

```sh
 busybox find --help              
BusyBox v1.27.2 (Ubuntu 1:1.27.2-2ubuntu3.2) multi-call binary.

Usage: find [-HL] [PATH]... [OPTIONS] [ACTIONS]

Search for files and perform actions on them.
First failed action stops processing of current file.
Defaults: PATH is current directory, action is '-print'

	-L,-follow	Follow symlinks
	-H		...on command line only
	-xdev		Don't descend directories on other filesystems
	-maxdepth N	Descend at most N levels. -maxdepth 0 applies
			actions to command line arguments only
	-mindepth N	Don't act on first N levels
	-depth		Act on directory *after* traversing it

Actions:
	( ACTIONS )	Group actions for -o / -a
	! ACT		Invert ACT's success/failure
	ACT1 [-a] ACT2	If ACT1 fails, stop, else do ACT2
	ACT1 -o ACT2	If ACT1 succeeds, stop, else do ACT2
			Note: -a has higher priority than -o
	-name PATTERN	Match file name (w/o directory name) to PATTERN
	-iname PATTERN	Case insensitive -name
	-path PATTERN	Match path to PATTERN
	-ipath PATTERN	Case insensitive -path
	-regex PATTERN	Match path to regex PATTERN
	-type X		File type is X (one of: f,d,l,b,c,...)
	-perm MASK	At least one mask bit (+MASK), all bits (-MASK),
			or exactly MASK bits are set in file's mode
	-mtime DAYS	mtime is greater than (+N), less than (-N),
			or exactly N days in the past
	-mmin MINS	mtime is greater than (+N), less than (-N),
			or exactly N minutes in the past
	-newer FILE	mtime is more recent than FILE's
	-inum N		File has inode number N
	-user NAME/ID	File is owned by given user
	-group NAME/ID	File is owned by given group
	-size N[bck]	File size is N (c:bytes,k:kbytes,b:512 bytes(def.))
			+/-N: file size is bigger/smaller than N
	-links N	Number of links is greater than (+N), less than (-N),
			or exactly N
	-prune		If current file is directory, don't descend into it
If none of the following actions is specified, -print is assumed
	-print		Print file name
	-print0		Print file name, NUL terminated
	-exec CMD ARG ;	Run CMD with all instances of {} replaced by
			file name. Fails if CMD exits with nonzero
	-exec CMD ARG + Run CMD with {} replaced by list of file names

File: ci/test/04_install.sh (hunk 1 above)
Comment (MarcoFalke) [reply ci/test/04_install.sh]: I have no idea what bionic is doing, but https://github.com/bitcoin/bitcoin/pull/17011#issuecomment-537507555

**Comment by MarcoFalke** [comment]
re run ci

**Comment by emilengler** [comment]
Concep ACK, in general it is better to rely on BusyBox rather GNU because GNU has features which other unix systems might not have.

**Comment by laanwj** [comment]
> Concep ACK, in general it is better to rely on BusyBox rather GNU because GNU has features which other unix systems might not have.

I guess it also reduces the number of packages that need to be built (are you using busybox utils in GUIX @dongcarl?).

Still, we want to run some builds with GNU utilities too, to make sure compatibility is not broken there.

**Comment by emilengler** [comment]
@laanwj Sure thing but everything which works under busybox will work under GNU but not necessarily vice versa.

**Comment by MarcoFalke** [comment]
@emilengler Is this documented somewhere?

**Comment by laanwj** [comment]
> @laanwj Sure thing but everything which works under busybox will work under GNU but not necessarily vice versa.

Even if theoretically true. I don't think any software can ever guarantee to be a strict subset of another.

**Comment by emilengler** [comment]
@MarcoFalke 
https://www.youtube.com/watch?v=nTCHapo8QFM

**Pull request merged**